initial omnibusf4sd target support

nathantsoi opened the pull request:
this PR is to track changes as development work is done on the omnibusf4sd target. it is not ready to merge


working:
  - ~~startup log on uart1, switches to mavlink on init~~, ~~moved to uart3~~, moved to uart4, uart1 is now serialrx
  - mpu6000, bench tested and verified via nsh
  - fmu
    - all 6 ch output bench tested w/ pwm and oneshot via nsh
    - ppm input bench tested
    - dsm input bench tested
  - bmp280, bench tested and verified via nsh
  - hmc5883, bench tested and verified via nsh, but requires an external i2c pullup
  - gps on uart6
  - nsh, mavlink on uart4. RSSI pin is TX, MOTOR 5 is RX (normal mode, 57600 baud)
  - rgbled, bench tested and working, but requires an external i2c pullup
  - sbus via the shared sbus/ppm pin (which includes an inverter to the mcu SBUS in pin), remove the solder bridge or jumper to the ppm pin before use

implemented, but not verified:
  - ~~hmc5883~~
  - ~~gps on uart6~~
  - ~~sbus~~

implemented, but not working:
  - ~~fmu~~
  - ~~bmp280~~
  - ~~uart3 -- "Fault log armed" is the last message on this uart, then no output from what should be the console then mavlink (i think)~~ -- cannot be used, SPI3 pins are in use

missing:
  - adc
  - osd, note: may need some basic driver to make sure this doesn't black out while flying
  - ~~rgbled control~~



~~bootloader pr is here: https://github.com/PX4/Bootloader/pull/101/files~~

~~tracking the general conversation and more detailed status information is here: http://discuss.px4.io/t/omnibusf4sd-target-support-request-for-development-guidance/6373~~

~~overall status is that this change against 1.7.4 is working, see the tree here: https://github.com/nathantsoi/Firmware/tree/omnibusf4sd-1.7.4~~

~~however, after rebasing against master, it does not build: https://gist.github.com/nathantsoi/557f04a48b3acb216617549d67cfb337~~

any guidance would be much appreciated

dagar commented:
Hi Nathan,

Try adding drivers/stm32 in your cmake config. https://github.com/PX4/Firmware/blob/master/cmake/configs/nuttx_px4fmu-v2_default.cmake#L46

Once this is working again we should get an initial version merged and building on Jenkins. In your PR add the target here - https://github.com/PX4/Firmware/blob/master/Jenkinsfile#L67

dagar commented:
@nathantsoi is this the correct board for this PR? Is there an all in one (AIO) option this should work with? I'm not up to speed on all these different betaflight boards.

### Omnibus Betaflight F4 PRO V2 Flight Controller STM32 F405
https://hobbyking.com/en_us/omnibus-betaflight-f4-pro-v2-flight-controller-stm32-f405.html

![image](https://user-images.githubusercontent.com/84712/38770495-e415d7be-3fe1-11e8-87ce-c0afc0c9c017.png)
![image](https://user-images.githubusercontent.com/84712/38770502-eccaa358-3fe1-11e8-9e65-2418e1dd2c3a.png)

nathantsoi commented:
yes, that board should work. i am testing with an original, available via Airbot: https://store.myairbot.com/omnibusf4prov3.html but any of the many clones should have the same pinout. i believe rtfq is a distributor near you: http://www.readytoflyquads.com/flip-32-f4-omnibus-v2-pro

i added back `drivers/stm32` per your suggestion and the hrt is fixed, but i'm still overflowing vs. master (but not 1.7.4) any ideas what i can remove / what changed?

```
/usr/local/Cellar/gcc-arm-none-eabi/20160928/bin/../lib/gcc/arm-none-eabi/5.4.1/../../../../arm-none-eabi/bin/ld: nuttx_px4-omnibusf4sd_default.elf section `.text' will not fit in region `flash'
/usr/local/Cellar/gcc-arm-none-eabi/20160928/bin/../lib/gcc/arm-none-eabi/5.4.1/../../../../arm-none-eabi/bin/ld: address 0x81041dc of nuttx_px4-omnibusf4sd_default.elf section `__param' is not within region `flash'
/usr/local/Cellar/gcc-arm-none-eabi/20160928/bin/../lib/gcc/arm-none-eabi/5.4.1/../../../../arm-none-eabi/bin/ld: address 0x81041dc of nuttx_px4-omnibusf4sd_default.elf section `__param' is not within region `flash'
/usr/local/Cellar/gcc-arm-none-eabi/20160928/bin/../lib/gcc/arm-none-eabi/5.4.1/../../../../arm-none-eabi/bin/ld: address 0x81041dc of nuttx_px4-omnibusf4sd_default.elf section `__param' is not within region `flash'
/usr/local/Cellar/gcc-arm-none-eabi/20160928/bin/../lib/gcc/arm-none-eabi/5.4.1/../../../../arm-none-eabi/bin/ld: region `flash' overflowed by 18721 bytes
```

dagar commented:
Thanks for the info.

For your build take a look at the included modules (cmake/configs/nuttx_px4-omnibusf4sd_default.cmake) and comment out what you definitely won't be using.

![image](https://user-images.githubusercontent.com/84712/38771533-d5b37ad0-3ff2-11e8-9443-70b3e58b72aa.png)
You can definitely comment out position_estimator_inav, and likely local_position_estimator + attitude_estimator_q as well.

nathantsoi commented:
sweet, thx @dagar. it builds!

dagar commented:
Great! Overall this PR is looking good (other than the python path stuff). The board name prefix (eg px4) is supposed to be for the vendor, and eventually I'd like to get all the board pieces centralized and sorted.

For example with px4fmu-v2 the vendor is px4 and the board name is fmu-v2.

Is there a vendor name that would make sense in place of px4 or would it be Omnibus with the board being f4sd or something?

    px4-omnibusf4sd -> omnibus-f4sd

nathantsoi commented:
thx again for all the feedback @dagar. i can rebase these commits after review, left them separate for now to ensure clarity on each change

- re. target naming, good to know. i've re-named the target to omnibus-f4sd 

- i added it to the jenkins file per your suggestion

- and removed the python code

the pwm app is not operating correctly, so i'm looking at the pwm and timer definitions, if you have any suggestions there? i need to look more closely at how the pin / timer / BOARD_FMU_GPIO_TAB definitions work together.

i also need to test ppm in and serial uart input as well as the gps / external mag.

dagar commented on platforms/nuttx/nuttx-configs/omnibus-f4sd/scripts/ld.script:
```diff
@@ -0,0 +1,150 @@
+/****************************************************************************
+ * configs/px4-stm32f4discovery/scripts/ld.script
```
Wrong path (not that it matters).

nathantsoi replied:
good catch

dagar commented on src/drivers/boards/omnibus-f4sd/spi.c:
```diff
@@ -0,0 +1,194 @@
+/****************************************************************************
+ *
+ *   Copyright (C) 2012 PX4 Development Team. All rights reserved.
```
Update

nathantsoi replied:
sure, just the date or should i add my name under the copyright?

dagar replied:
That's up to you. Personally I use PX4 Development Team, but you're welcome to use your name if you'd prefer. Either way with the current date.

nathantsoi replied:
on it. thx!

hamishwillee commented:
@nathantsoi When this is merged, do you think we could get some documentation in the PX4 user guide? 

Is this a px4fmu_v2 target? (or v3?). 

Minimally we'd have a page for the board like [mro](https://docs.px4.io/en/flight_controller/mro_x2.1.html), [HKpilot32](https://docs.px4.io/en/flight_controller/HKPilot32.html) etc, covering:
* High level overview - marketing paragraph about the board
* Quick summary - sensors, CPU, memory, etc
* Availability - link to store, accessories
* Wiring guides
* Links to schematics and pinouts

**Ideally** we'd have a wiring quickstart [like this one](https://docs.px4.io/en/assembly/quick_start_pixhawk_mini.html) which shows how to connect up the board to all the major components and power for multicopter and ideally also planes. Some vendors have that information on their own site and just link to it.

bkueng commented on ROMFS/px4fmu_common/init.d/rcS:
```diff
@@ -403,12 +414,17 @@ then
+		fi
+		if ver hwcmp OMNIBUS_F4SD
+		then
+			set GPS_OPT "-d /dev/ttyS1"
```
Can you instead add a line `#define GPS_DEFAULT_UART_PORT  "/dev/ttyS1"      /* UART6 */` to the board_config.h file?

dagar commented:
@nathantsoi there's actually quite a bit of interest in DSHOT support (surprising to me at least). https://github.com/PX4/Firmware/issues/8182

You might want to pursue it in a separate PR. Feel free to open a new issue if you'd like to discuss the details. I suspect we'll need to figure out some board level support for DMA deconfliction.

nathantsoi commented:
ahh, sorry, I mistyped above. `dshot` should be `oneshot`. though, dshot is a great feature and I'd love to contribute this at some point, i don't plan on doing dshot in this pr. just want to get the timers and io working ofr now

as an aside dma for oneshot in betaflight is (4ch only):
```
// PB0 S1_OUT D1_ST7 -- that is DMA1 Stream7
// PB1 S2_OUT D1_ST2
// PA3 S3_OUT D1_ST6
// PA2 S4_OUT D1_ST1
```

dagar commented:
> i don't plan on doing dshot in this pr. just want to get the timers and io working for now

That makes more sense, I wasn't sure how ambitious you are. :smile: 

> the pwm app is not operating correctly, so i'm looking at the pwm and timer definitions, if you have any suggestions there? i need to look more closely at how the pin / timer / BOARD_FMU_GPIO_TAB definitions work together.

So starting from the beginning are the timers in the board_config.h and timer_config.c set appropriately as far as you can tell? How does it fail? Does the fmu module start? Does the pwm command report anything?

    fmu status
    pwm info

dagar commented:
Oh, and take a look at the CI failure. http://ci.px4.io:8080/blue/organizations/jenkins/Firmware/detail/PR-9310/6/pipeline

It's a trivial style check failure. You can run it locally first if you install astyle and `make check_format`.

hamishwillee commented:
@nathantsoi 

> happy to add docs @hamishwillee, i would like to get pwm out, dshot oneshot, ppm in and serial rx working first.

That is excellent. Yes, best if board is "solid" for use with PX4 before we add user guide documentation. I have added issue to track that here: https://github.com/PX4/px4_user_guide/issues/200
Instructions on the docs toolchain are [here](https://dev.px4.io/en/contribute/docs.html) - but feel free to ping me for questions.

bkueng commented:
> now i need to verify ppm is working. any tips on this? i just started looking at drivers/test_ppm, i assume this should help me verify the input is working?


`test_ppm` could help, but I've never used it. The best way to test would be to attach an actual PPM receiver and check the channels (and `fmu status`: `RC scan state` should switch to PPM and stay there). Do you have one? Some receivers can be put into PPM mode by using a jumper.

nathantsoi commented:
thx for the tips @bkueng, ppm is working!

i was seeing this in nsh after disabling the rc serial port:

```
nsh> fmu status
INFO  [fmu] Running as task
INFO  [fmu] RC scan state: SBUS
INFO  [fmu] PWM Mode: pwm6
```

tracing through the ppm code though it seems like it was working. plugged it into qgroundcontrol, ppm is working great!

i had to disable `RC_SERIAL_PORT` to get it working though. any tips on where to start getting serial rx protocols (SBUS, DSUM, etc.) running as well (via uart1) would be appreciated

also, any tips on the right place to specify board-specific imu orientation would be great... qground control thinks it's upside down.

bkueng commented:
> thx for the tips @bkueng, ppm is working!

Cool!

> i had to disable RC_SERIAL_PORT to get it working though. any tips on where to start getting serial rx protocols (SBUS, DSUM, etc.) running as well (via uart1) would be appreciated

What you should be able to do is to define `RC_SERIAL_PORT` to the UART port that you want and then the `fmu` driver will auto-detect the protocol.

> also, any tips on the right place to specify board-specific imu orientation would be great... qground control thinks it's upside down.

You can do that in the startup script (https://github.com/PX4/Firmware/pull/9310/files#diff-dc146e6fb305825fdba36bf461a3e970R272), by adding a rotation parameter in the form `-R <integer rotation>` to the driver startup. The rotations are defined here: https://github.com/PX4/Firmware/blob/master/src/lib/conversion/rotation.h#L50

nathantsoi commented:
quick update, the following are now working:
- ppm (confirmed on the scope, not on ESCs, yet)
- gyro orientation is correct now (thx @bkueng)
- gps over UART6 is working (e.g. here):
![screen shot 2018-04-22 at 12 14 10 am](https://user-images.githubusercontent.com/399770/39092414-1c95e7ce-45c2-11e8-8487-b646d2642489.png)

the last thing I'd like to get working for this PR is the external I2C2 bus (J10 SCL/SCA in the above photo). I have tested both rgbled and hmc5883, but neither are working. maybe I'm missing something obvious?

tracing https://github.com/PX4/Firmware/blob/master/src/drivers/magnetometer/hmc5883/hmc5883_i2c.cpp#L133-L138, I see:
![screen shot 2018-04-22 at 12 20 22 am](https://user-images.githubusercontent.com/399770/39092447-fe50d502-45c2-11e8-85db-f85239b5385f.png)
![screen shot 2018-04-22 at 12 20 34 am](https://user-images.githubusercontent.com/399770/39092449-01bd6868-45c3-11e8-99b0-1d5f7123360b.png)

and in nsh:
```
nsh> hmc5883 info
ERROR [hmc5883] bus 0 not started
nsh> hmc5883 -X info
ERROR [hmc5883] bus 2 not started
```

I thought it was maybe the slave, so I wired up an rgbled i2c breakout board:

```
nsh> rgbled start
WARN  [rgbled] no RGB led on bus #2
```

with just the rgbled attached, the trace is almost the same, full trace from board startup:
![screen shot 2018-04-22 at 12 26 07 am](https://user-images.githubusercontent.com/399770/39092501-fc8c0556-45c3-11e8-95cc-69a5737824a4.png)

![screen shot 2018-04-22 at 12 26 41 am](https://user-images.githubusercontent.com/399770/39092497-e0fb08f0-45c3-11e8-9ebc-dd70866c3cc0.png)

nathantsoi commented:
awesome! the rgbled and the hmc5883 both work on the external bus with w/ 2.2k pullups on the clock and data lines... now is it best to suggest users add these pullups externally (I wired vcc from the dsm2 port), which seems more robust. or should i try to fix this in software via internal pullup on the gpio, if possible?

```
nsh> hmc5883 info
WARN  [hmc5883] running on bus: 2 (/dev/hmc5883_ext)

hmc5883_read: 2520 events, 2503767us elapsed, 993us avg, min 867us max 6312us 263.842us rms
hmc5883_com_err: 0 events
poll interval:  7 ticks
 sensor_mag_s
	timestamp: 23190598 (0.001623 seconds ago)
	error_count: 0
	x: 0.398
	y: 0.290
	z: 0.407
	range_ga: 1.900
	scaling: 0.001
	temperature: 0.000
	device_id: 73233
	x_raw: 326
	y_raw: 238
	z_raw: 334
	is_external: 1
report queue	2/112 (0/2 @ 10009510)
```

LorenzMeier commented:
Sounds like an attempt on the GPIO would be worthwhile

nathantsoi commented:
Ok, I enabled the pull-up (I think, see this patch: https://gist.github.com/nathantsoi/bcf3a99402ef0287eb576df60772df2c), however the pull-up appears to weak to pull the rgbled (and debugging) setup up to 3v. I only see 2v on the scope:

![img_20180423_215315](https://user-images.githubusercontent.com/399770/39167093-1f76c96e-4742-11e8-8146-9a801a115241.jpg)

I also wonder if the power-up-low-state of the bus is problematic.

For reference, here is a trace with 2.2k pullups:

![img_20180423_214927](https://user-images.githubusercontent.com/399770/39167113-3c9b187e-4742-11e8-9b2d-1aeeb66cd0f6.jpg)

I'm open to ideas, but I wonder if we'll need to add the external pullups?

LorenzMeier commented:
Sounds like an external pullup is required.

nathantsoi commented:
sounds good @LorenzMeier. I may try to bring the i2c lines up sooner and see if it helps, but will stick for the external pullups for now.

this is getting much closer. I hope to flight test on a small quad soon.

@dagar, any tips on fixing this CI failure? I assume this is not an existing issue
```
FAILED: : && /usr/lib/ccache/arm-none-eabi-g++   -fno-common -ffunction-sections -fdata-sections -mcpu=cortex-m4 -mthumb -march=armv7e-m -mfpu=fpv4-sp-d16 -mfloat-abi=hard -fno-common -ffunction-sections -fdata-sections -g -fno-exceptions -fno-rtti -std=gnu++11 -fno-threadsafe-statics -DCONFIG_WCHAR_BUILTIN -D__CUSTOM_FILE_IO__ -fcheck-new -Wall -Warray-bounds -Wdisabled-optimization -Werror -Wextra -Wfatal-errors -Wfloat-equal -Wformat-security -Winit-self -Wlogical-op -Wmissing-declarations -Wmissing-field-initializers -Wpointer-arith -Wshadow -Wuninitialized -Wunknown-pragmas -Wunused-variable -Wno-implicit-fallthrough -Wno-unused-parameter -Wunused-but-set-variable -Wformat=1 -Wdouble-promotion -Wno-missing-field-initializers -Wno-overloaded-virtual -Wreorder -fvisibility=hidden -include visibility.h -fno-strict-aliasing -fomit-frame-pointer -fno-math-errno -funsafe-math-optimizations -ffunction-sections -fdata-sections -fno-strength-reduce -fno-builtin-printf -Os -DNDEBUG  -mcpu=cortex-m4 -mthumb -march=armv7e-m -mfpu=fpv4-sp-d16 -mfloat-abi=hard -nodefaultlibs -nostdlib -Wl,--warn-common,--gc-sections platforms/nuttx/CMakeFiles/nuttx_px4fmu-v2_default.elf.dir/__/__/src/platforms/empty.c.obj  -o nuttx_px4fmu-v2_default.elf -L/tmp/jenkins/workspace/Firmware_PR-9310-P3AQN4QTSTHGWXXZV7QP4H7ZD2N54XOVWX6IGXWKUARALGSIPVJA/build/nuttx_px4fmu-v2_default/external/Install/lib -Wl,--script=/tmp/jenkins/workspace/Firmware_PR-9310-P3AQN4QTSTHGWXXZV7QP4H7ZD2N54XOVWX6IGXWKUARALGSIPVJA/build/nuttx_px4fmu-v2_default/NuttX/nuttx/configs/px4fmu-v2/scripts/ld.script -Wl,-Map=nuttx_px4fmu-v2_default.map -Wl,--warn-common,--gc-sections -Wl,--start-group msg/libuorb_msgs.a src/platforms/common/libplatforms__common.a src/drivers/differential_pressure/ets/libdrivers__ets_airspeed.a src/drivers/differential_pressure/ms4525/libdrivers__ms4525_airspeed.a src/drivers/differential_pressure/ms5525/libdrivers__ms5525_airspeed.a src/drivers/differential_pressure/sdp3x/libdrivers__sdp3x_airspeed.a src/drivers/airspeed/libdrivers__airspeed.a src/drivers/barometer/ms5611/libdrivers__ms5611.a src/drivers/boards/common/libboards_common.a src/drivers/boards/common/stm32/libboard_stm32.a src/drivers/boards/px4fmu-v2/libdrivers_board.a src/drivers/camera_trigger/libdrivers__camera_trigger.a src/drivers/device/libdrivers__device.a src/drivers/gps/libdrivers__gps.a src/drivers/imu/l3gd20/libdrivers__l3gd20.a src/drivers/led/libdrivers__led.a src/drivers/imu/lsm303d/libdrivers__lsm303d.a src/drivers/magnetometer/hmc5883/libdrivers__hmc5883.a src/drivers/magnetometer/lis3mdl/libdrivers__lis3mdl.a src/drivers/imu/mpu6000/libdrivers__mpu6000.a src/drivers/imu/mpu9250/libdrivers__mpu9250.a src/drivers/pwm_input/libdrivers__pwm_input.a src/drivers/px4flow/libdrivers__px4flow.a src/drivers/px4fmu/libdrivers__px4fmu.a src/drivers/px4io/libdrivers__px4io.a src/drivers/rgbled/libdrivers__rgbled.a src/drivers/stm32/libdrivers__stm32.a src/drivers/stm32/adc/libdrivers__stm32__adc.a src/drivers/stm32/tone_alarm/libdrivers__stm32__tone_alarm.a src/drivers/vmount/libdrivers__vmount.a src/drivers/distance_sensor/ll40ls/libdrivers__ll40ls.a src/drivers/distance_sensor/mb12xx/libdrivers__mb12xx.a src/drivers/distance_sensor/sf0x/libdrivers__sf0x.a src/drivers/distance_sensor/sf1xx/libdrivers__sf1xx.a src/drivers/distance_sensor/srf02/libdrivers__srf02.a src/drivers/distance_sensor/srf02_i2c/libdrivers__srf02_i2c.a src/drivers/distance_sensor/teraranger/libdrivers__teraranger.a src/drivers/distance_sensor/tfmini/libdrivers__tfmini.a src/modules/sensors/libmodules__sensors.a src/systemcmds/hardfault_log/libsystemcmds__hardfault_log.a src/systemcmds/mixer/libsystemcmds__mixer.a src/systemcmds/mtd/libsystemcmds__mtd.a src/systemcmds/param/libsystemcmds__param.a src/systemcmds/perf/libsystemcmds__perf.a src/systemcmds/pwm/libsystemcmds__pwm.a src/systemcmds/reboot/libsystemcmds__reboot.a src/systemcmds/top/libsystemcmds__top.a src/systemcmds/tune_control/libsystemcmds__tune_control.a src/systemcmds/ver/libsystemcmds__ver.a src/modules/camera_feedback/libmodules__camera_feedback.a src/modules/commander/libmodules__commander.a src/modules/events/libmodules__events.a src/modules/land_detector/libmodules__land_detector.a src/modules/load_mon/libmodules__load_mon.a src/modules/mavlink/libmodules__mavlink.a src/modules/navigator/libmodules__navigator.a src/modules/ekf2/libmodules__ekf2.a src/modules/fw_att_control/libmodules__fw_att_control.a src/modules/fw_pos_control_l1/launchdetection/liblaunchdetection.a src/modules/fw_pos_control_l1/runway_takeoff/librunway_takeoff.a src/modules/fw_pos_control_l1/libmodules__fw_pos_control_l1.a src/modules/mc_att_control/libmodules__mc_att_control.a src/modules/mc_pos_control/libmodules__mc_pos_control.a src/modules/vtol_att_control/libmodules__vtol_att_control.a src/modules/logger/libmodules__logger.a src/modules/dataman/libmodules__dataman.a src/modules/systemlib/libmodules__systemlib.a src/modules/systemlib/param/libmodules__systemlib__param.a src/modules/uORB/libmodules__uORB.a src/lib/controllib/liblib__controllib.a src/lib/conversion/liblib__conversion.a src/lib/DriverFramework/framework/libDriverFramework__framework.a src/lib/ecl/libecl.a src/lib/led/liblib__led.a src/lib/mathlib/liblib__mathlib.a src/lib/mixer/liblib__mixer.a src/lib/tunes/liblib__tunes.a src/lib/version/liblib__version.a src/lib/FlightTasks/liblib__flight_tasks.a platforms/nuttx/src/px4_layer/libplatform_px4_layer.a NuttX/apps/libapps.a NuttX/nuttx/arch/arm/src/libarch.a NuttX/nuttx/binfmt/libbinfmt.a NuttX/nuttx/binfmt/libbinfmt.a NuttX/nuttx/libc/libc.a NuttX/nuttx/configs/libconfigs.a NuttX/nuttx/libxx/libcxx.a NuttX/nuttx/drivers/libdrivers.a NuttX/nuttx/fs/libfs.a NuttX/nuttx/mm/libmm.a NuttX/nuttx/sched/libsched.a -Wl,--end-group -lm -lgcc ROMFS/libromfs.a src/lib/DriverFramework/framework/src/libdf_driver_framework.a && :
```

nathantsoi commented:
@bkueng, i merged https://github.com/PX4/Firmware/pull/9361 with this branch and bench tested the code here: https://github.com/nathantsoi/Firmware/tree/omnibusf4sd-q_estimator_improvements

i had to skip the mag in preflight checks: https://github.com/nathantsoi/Firmware/commit/0d6249ad412c99f58b36e94bf793eb52a1ef90b0, lmk where this should go and i can open a PR

i also made it possible to configure flight modes w/o a mag here: https://github.com/mavlink/qgroundcontrol/pull/6383

finally, bench testing stabilized mode with the generic 250 race frame and this motor layout https://dev.px4.io/en/airframes/airframe_reference.html#quadrotor-x, i get exactly the opposite motor response as I would expect.

e.g. tip the quad forward -> forward motors spool up (I would expect the back motors to spool up)
tip right -> right motors spool up, etc.

thoughts on where i might correct this?

bkueng commented:
Cool :)

I think we a more generic way for systems without a mag, rather than using `ATT_W_MAG`. I see these options:
- Add a parameter `SYS_HAS_MAG` that is respected for the preflight checks and QGC (if disabled, it would imply `ATT_W_MAG=0`)
- Make the mag optional in general and for preflight checks (only require that if a mag is found, it needs to be calibrated). We can let QGC know through [SYS_STATUS](https://mavlink.io/en/messages/common.html#SYS_STATUS) or some other way if the vehicle has a mag.
That's my preferred solution because it does not add another parameter.

@dagar @DonLakeFlyer @LorenzMeier  What's your take on this?

---

> e.g. tip the quad forward -> forward motors spool up (I would expect the back motors to spool up)
tip right -> right motors spool up, etc.

Am I reading it wrong or isn't that what you would expect? Are you testing in Stabilized/Manual? Tipping it forward would make the front motors spin up, because it tries to go back to level.

LorenzMeier commented:
This definitely needs a parameter. A mag is definitely part of the minimum equipment list and no, "showing it to the user" never works - how would they know that they need a mag for position control flight mode?

Racers are an "extreme" case of the stack and will always require configuration parameters to configure it accordingly. We also will need to configure EKF and many other system params to make the system behave for racers without spoiling the reliability and hardening we have for our main use cases.

nathantsoi commented:
thanks guys!

you're right @bkueng, from how i described, it that would make sense. i guess it would better to say the quad was trying to do the opposite of level :) i was worried about motor order, mixer, etc. turns out i had the fc oriented 180 degrees the wrong way. e.g. i just wired it wrong. it's fine now that i wired it correctly :)

i did try to hover a quad today, didn't make it off the ground though. from first glance it looks like there is just wayy to much noise on the gyro https://logs.px4.io/plot_app?log=ff5aa81d-acf6-4315-b77b-40a93f8b7705 (i assume this is the log analysis tool i should be using?) any suggestions on filtering?

let me know how i should proceed for mag-less arming. happy to help implement whatever is required here

DonLakeFlyer commented:
Similar concept is handled for airspeed sensor with QGC checking both FW_ARSP_MODE and CBRK_AIRSPD_CHK to determine whether airspeed needs calibration. Doing it with SYS_STATUS would be different from other calibration checks in QGC which are parameter based.

dagar commented:
I think we need system wide parameters for the vehicle configuration. 

It would also be better for a ground control station to get this information via a mavlink stream rather than parameters.

DonLakeFlyer commented:
> It would also be better for a ground control station to get this information via a mavlink stream rather than parameters.

I'm not entirely convinced of that. A parameter already has the concept of a protocol to get data across a noisy link as well as that protocol supporting update notifications. Not to mention meta data and so forth as well. I think there may be something missing which is in between a mavlink message sent to QGC and a parameter that would be useful to get status information from.

dagar commented:
@DonLakeFlyer yes that's obviously better in just about every way other than not being defined in a spec.

bkueng commented:
Let's go for a `SYS_HAS_MAG` param then?

@nathantsoi Can you enable the high rate logging in `SDLOG_PROFILE` (set it to 19)? It will help you with rate tuning. Try to start off with considerably lower rate D & P gains (e.g. MC_ROLLRATE_P=0.08 and MC_ROLLRATE_D=0.001).

nathantsoi commented:
I actually see the same thing on USART3, @bkueng. I am just starting to debug this.

The only output on this USART is:
```
[boot] Fault Log info File No 4 Length 3177 flags:
```
Does that mean anything to anyone?

This makes me think something else is taking control of the port or there is some hardware conflict? Also, maybe related, I have been unable to get NSH working over serial, even on a working U(S)ART (e.g. USART6).

Re. `SYS_HAS_MAG`, sounds good. For now I put an external mag/compass on my test quad so I can test with all estimators without significant (removing mag requirement) changes. Since testing without the mag is such a big change (I did not realize this before) I will test w/o the mag after the initial port.

Re. `SDLOG_PROFILE`, thank you! I will turn this up. Also, I did not realize the default gains were not low enough. I will bring those down to the values you suggest. Finally, it looks like https://github.com/PX4/Firmware/pull/7446 may help with sampling speed and present an opportunity for lowpass / notch filtering, if these do not exist already?

nathantsoi commented:
Also, re. i2c on the omnibus boards, here is some more info on external pullups: https://github.com/iNavFlight/inav/issues/1877

While I did revisit this, trying to bring up the internal pull-ups sooner, it is not enough. We will need these externally.

I built a little cable that takes 3.3v from the Spektrum port and attaches to the I2C SDA/SDL lines via 2.2k resistor (one resist per line).

![screen shot 2018-05-02 at 8 56 34 am](https://user-images.githubusercontent.com/399770/39534624-d8b136bc-4de6-11e8-971c-785d6b0e127f.png)

bkueng commented:
> Does that mean anything to anyone?

Yes, it's the first output on the debug shell, coming from https://github.com/PX4/Firmware/blob/master/ROMFS/px4fmu_common/init.d/rcS#L63, due to `CONFIG_USART3_SERIAL_CONSOLE=y`. You should get the shell there.

> Also, I did not realize the default gains were not low enough. I will bring those down to the values you suggest.

People have traditionally not used powerful small vehicles with PX4 - we will have to change the defaults.

> Finally, it looks like #7446 may help with sampling speed and present an opportunity for lowpass / notch filtering, if these do not exist already?

Precisely. I see disabling the onchip filter & increasing sampling rate as one of the next steps to improve flight performance. But it seems we will need DMA to do that. Our current filter pipeline for the controllers is:
- on-chip filter at 96Hz sampled with 1kHz
- SW low-pass filter & downsample to 250Hz (`IMU_GYRO_CUTOFF` param)
- D-term low-pass filter (`MC_DTERM_CUTOFF` param)
- no notch filter yet, but it's also something I'd like to see being added/tested

nathantsoi commented:
thx @bkueng. i figured out what's up with uart3, it's on the SPI3 pins, which are in use by the onboard bmp280 and max4756.

so i have added uart4 instead on the RSSI pin (TX) PWM out 5 (RX): https://github.com/PX4/Firmware/pull/9310/commits/8f88022b36bad54530115115e7e1935912a83b92#diff-cf85c24d6c752f3e96fcda174e5a4f1dR275

![screen shot 2018-05-03 at 11 17 53 pm](https://user-images.githubusercontent.com/399770/39614931-47e7cb8a-4f28-11e8-8b4b-cf7255dc43b4.png)

![screen shot 2018-05-03 at 11 18 02 pm](https://user-images.githubusercontent.com/399770/39614936-4a9aab18-4f28-11e8-8aa3-88f0ab49263a.png)

i also got the NSH working on uart4, so if an `etc/rc.txt` file is present on the sdcard, one can get to `nsh`

nathantsoi commented:
![screen shot 2018-05-04 at 7 49 50 pm](https://user-images.githubusercontent.com/399770/39667381-ce07e082-5069-11e8-9a6a-94d0b143a4d1.png)

this is what i see when connected to the hc-06

nathantsoi commented:
Ok, update! I flew a 280mm quad w/ omnibusf4sd, mag, and gps: https://logs.px4.io/plot_app?log=79b81031-cf1e-41f0-890b-d6cd7d559766

I had to turn down the default pids quite a bit, as you suggested @bkueng, but with almost no tuning, it flew well!

I think this is getting close to being merge-able. I will look at the battery ADC and then rebase everything.

bkueng commented:
Cool! :)

> this is what i see when connected to the hc-06

It typically means that the connection is only one-way, i.e. the vehicle does not receive anything. Is the [mavlink shell](https://github.com/mavlink/qgroundcontrol/pull/4913) working? `mavlink status` tells you if the vehicle receives data.

bkueng commented:
@nathantsoi is the motor assignment correct? With the 4-in-1 esc I have:

```
4     2
   ^
3     1
```
Whereas it should be:
```
3     1
   ^
2     4
```

nathantsoi commented:
I think it is correct... I will double check my wiring tonight. What 4-in-1
are you using?

On Mon, May 7, 2018 at 8:05 AM Beat Küng <notifications@github.com> wrote:

> @nathantsoi <https://github.com/nathantsoi> is the motor assignment
> correct? With the 4-in-1 esc I have:
>
> 4     2
>    ^
> 3     1
>
> Whereas it should be:
>
> 3     1
>    ^
> 2     4
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PX4/Firmware/pull/9310#issuecomment-387095171>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAYZmmY4D1jQzWvaQoUb8O5opmY8lXZKks5twGK5gaJpZM4TU95k>
> .
>

bkueng commented:
This one: http://www.hobbywing.com/goods.php?id=588. You can just plug the Omnibus on top of it.

nathantsoi commented:
I double-checked the wiring on my test-flight quad @bkueng and I see that the motor order does match the omnibus f4 pinout (note I am currently flight testing with an F4SD, but have an F4SD and an F4SD Pro on the bench, but have not noticed any discrepancies up to this point)

## OmnibusF4SD
![screen shot 2018-05-08 at 9 27 24 pm](https://user-images.githubusercontent.com/399770/39795708-ebf35316-5306-11e8-8f22-de9c36ab7064.png)
## PX4 Motor Mapping
![screen shot 2018-05-08 at 9 27 28 pm](https://user-images.githubusercontent.com/399770/39795709-ec46771c-5306-11e8-927c-c809608d9ba1.png)




Looks like the hobbywing esc linked above has a BetaFlight pin/motor mapping:

## BetaFlight motor mapping
![screen shot 2018-05-08 at 9 33 29 pm](https://user-images.githubusercontent.com/399770/39795789-8e9a3648-5307-11e8-897f-0b005a40fce3.png)
## Hobbywing BLHeliSDS600 Pin/Motor output
![screen shot 2018-05-08 at 9 30 45 pm](https://user-images.githubusercontent.com/399770/39795790-8ed168b6-5307-11e8-9cd0-b54ad7bebfbc.png)
per http://www.hobbywing.com/products/enpdf/BLHeliSDS600.pdf

bkueng commented:
You're right, from that standpoint everything is correct. But I see 2 problems:
- switching between Betaflight & PX4 requires physical changes of the setup
- the 4-in-1 ESC is not connected via JST (which would indeed be easy to swap), but directly via the pin header on the right:

![39795790-8ed168b6-5307-11e8-9cd0-b54ad7bebfbc](https://user-images.githubusercontent.com/281593/39803184-8ae44e12-5370-11e8-9674-328474dd87ed.png)

Maybe we should add a parameter that allows to use the Betaflight motor assignment?

bkueng commented:
I have it flying as well now :)
![img_20180509_201843293_ll](https://user-images.githubusercontent.com/281593/39916376-95e050cc-550a-11e8-9d03-fab90951367d.jpg)
Some notes:
- I changed the motor assignment in the `fmu` driver
- I used uart6 (gps) to attach the ESP8266 for mavlink, because I had similar issues on uart4 as you have (data seems to go through in both directions, but QGC fails to load all parameters)

---
I have also added the SYS_HAS_MAG param in https://github.com/bkueng/Firmware/commit/20978e5292c2d7359c24f365e175891a420ef6f7, https://github.com/bkueng/Firmware/commit/656e00edc9d57bdc939d69a50a5287b87eee4776 and https://github.com/bkueng/Firmware/commit/67ba173a5814ae6b9f54d948abfdc995c55c3ec7. Can you rebase on master and cherry-pick these?

nathantsoi commented:
Sweet! I wonder what's up w/ uart4...

The motor assignment in the fmu driver -- is that code or configuration? Do
we still need the motor order param?

Thx for the SYS_HAS_MAG code! I'll cherry-pick/rebase/squash tonight.

On Fri, May 11, 2018, 2:12 AM Beat Küng <notifications@github.com> wrote:

> I have it flying as well now :)
> [image: img_20180509_201843293_ll]
> <https://user-images.githubusercontent.com/281593/39916376-95e050cc-550a-11e8-9d03-fab90951367d.jpg>
> Some notes:
>
>    - I changed the motor assignment in the fmu driver
>    - I used uart6 (gps) to attach the ESP8266 for mavlink, because I had
>    similar issues on uart4 as you have (data seems to go through in both
>    directions, but QGC fails to load all parameters)
>
> ------------------------------
>
> I have also added the SYS_HAS_MAG param in bkueng/Firmware@20978e5
> <https://github.com/bkueng/Firmware/commit/20978e5292c2d7359c24f365e175891a420ef6f7>,
> bkueng/Firmware@656e00e
> <https://github.com/bkueng/Firmware/commit/656e00edc9d57bdc939d69a50a5287b87eee4776>
> and bkueng/Firmware@67ba173
> <https://github.com/bkueng/Firmware/commit/67ba173a5814ae6b9f54d948abfdc995c55c3ec7>.
> Can you rebase on master and cherry-pick these?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PX4/Firmware/pull/9310#issuecomment-388308370>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAYZmgmZg-suOnAbWrsEwRAsmZIKlHnKks5txVYWgaJpZM4TU95k>
> .
>

bkueng commented:
>  The motor assignment in the fmu driver -- is that code or configuration? Do we still need the motor order param? 

It's code and I think we need the param. Here's my diff:
```diff
diff --git a/src/drivers/px4fmu/fmu.cpp b/src/drivers/px4fmu/fmu.cpp
index 2434cbb366..2b9ff00fbb 100644
--- a/src/drivers/px4fmu/fmu.cpp
+++ b/src/drivers/px4fmu/fmu.cpp
@@ -1326,6 +1326,13 @@ PX4FMU::cycle()
                     }
                 }

+                const uint16_t pwm_tmp[4] = {pwm_limited[0], pwm_limited[1], pwm_limited[2], pwm_limited[3] };
+                pwm_limited[0] = pwm_tmp[3];
+                pwm_limited[1] = pwm_tmp[0];
+                pwm_limited[2] = pwm_tmp[1];
+                pwm_limited[3] = pwm_tmp[2];
+
+
                 /* output to the servos */
                 if (_pwm_initialized) {
                     for (size_t i = 0; i < mixed_num_outputs; i++) {
```

nathantsoi commented:
Rebased master and cherry-picked your commits @bkueng, thx for those

Also, I squashed everything, note that individual commits are preserved here for those interested in the progression of this PR: https://github.com/PX4/Firmware/compare/master...nathantsoi:omnibusf4sd-wip

The two remaining items are the ADC and OSD driver, but since I am slightly more time-constrained at the moment, perhaps we can merge this and work on those next. This is ready for review @dagar and @bkueng.

Also, @LorenzMeier mentioned that the EKF2 estimator may need some changes. I think mag-less mode and also perhaps some handling of quick-switching in and out of acro mode?

bkueng commented on ROMFS/px4fmu_common/init.d/rcS:
```diff
@@ -13,6 +13,12 @@ set +e
+# UART mapping on OMNIBUSF4SD:
+#
+# USART1		/dev/ttyS0		SerialRX
+# USART3		/dev/ttyS1		TELEM1
```
This is disabled now, correct?

nathantsoi replied:
yep, thx, switch to uart4

bkueng commented on ROMFS/px4fmu_common/init.d/rcS:
```diff
@@ -707,7 +728,7 @@ then
 	fi
 
-	if ver hwcmp PX4FMU_V2 PX4FMU_V4 PX4FMU_V4PRO MINDPX_V2 PX4FMU_V5
+	if ver hwcmp PX4FMU_V2 PX4FMU_V4 PX4FMU_V4PRO MINDPX_V2 PX4FMU_V5 OMNIBUS_F4SD
```
Is this on purpose? Since the px4flow driver is not enabled in the cmake file.

nathantsoi replied:
ahh, yes. thx. brought back px4flow in the cmake config

bkueng commented:
> Also, @LorenzMeier mentioned that the EKF2 estimator may need some changes. I think mag-less mode and also perhaps some handling of quick-switching in and out of acro mode?

Yes the goal is to make it work for such setups as well: https://github.com/PX4/Firmware/issues/9449.

davids5 commented:
> I have 2 questions, otherwise should be good to go.
@davids5 did you have a chance to review the board & NuttX config?

@bkueng - I would need the schematic, to be able to vet the config.

nathantsoi commented:
@davids5, pinouts should be interchangeable between the omnibusf4 pro (includes a current sensor) and the omnibusf4sd (no current sensor). Schematics are complements of Airbot (myairbot.com): [OmnibusF4-Pro-Sch.pdf](http://bit.ly/obf4pro)

nathantsoi commented:
@bkueng, thx for the review. i just squashed in this commit: https://github.com/nathantsoi/Firmware/commit/c5cacdd20106b5155f9a389fdd06b1f4bbdf6457

davids5 commented:
@nathantsoi - thank you for the schematic.  You did a great job on the bord config. What are the issues you are having currently? Specifically what is the issue you are having with UART4? Is it related to trying to used the port for console and mavlink?

nathantsoi commented:
@davids5, yea for some reason (haven't debugged too much yet) uart4 won't sent params to qgroundcontrol. I have a bluetooth module on this port and @bkueng has a esp8266. Same over uart/serial. Flight mode comes over fine. No idea if this is board-config related or something else. Thx again for reviewing this!

davids5 commented:
@nathantsoi are you building with the console disabled on UART4 and then using mavlink on it?

nathantsoi commented:
interesting @davids5, you're right. if i switch to `CONFIG_USART6_SERIAL_CONSOLE`, uart4 works fine.

i tried building w/ various combinations of `CONFIG_NO_SERIAL_CONSOLE=y`, even disabling syslog e.g. https://gist.github.com/nathantsoi/59d14fb300bd9a4f52b8e59705641c9f to no avail, I don't get mavlink on the USB port when I do this... how do I disable the console?

also, is it possible to run kconfig against the defconfig file in the nuttx-configs folder? maybe a better question is, should I be doing that at all? i've just been editing the config by hand and re-building

davids5 commented:
@nathantsoi - 

Because of the dependencies we should never edit defconfig files by hand. The correct thing to do is run `make <target> menuconfig` then revert the lines it should not have removed. It is obvious, which they are and the build tool tries to correct it but it may not be correct. These delete happen is because the PX4 targets are not in the Kconfig of the toplevel config directory in nuttx.

I would expect 

With console
```
CONFIG_UART4_SERIAL_CONSOLE=y
CONFIG_NO_SERIAL_CONSOLE is not set
```
Without console
```
CONFIG_UART4_SERIAL_CONSOLE=n
CONFIG_NO_SERIAL_CONSOLE=y
```

dagar commented on src/modules/sensors/parameters.cpp:
```diff
@@ -188,6 +188,9 @@ void initialize_parameter_handles(ParameterHandles &parameter_handles)
 	(void)param_find("SYS_CAL_TMAX");
 	(void)param_find("SYS_CAL_TMIN");
+
+	(void)param_find("SYS_HAS_MAG");
```
This isn't necessary.

nathantsoi replied:
i think i invalidated @bkueng's review/approval by squashing in his requested changes. i'll remove this line. @dagar -- want me to squash that in? separate commit or fix later?

nathantsoi replied:
never mind, i just pushed it separately since it's not in my original commit anyway

dagar replied:
It's pretty harmless to leave if you aren't already rebasing.

dagar commented:
There are some minor things, but from a PX4 perspective this should be safe to merge. Unfortunately Jenkins won't respect your Jenkinsfile changes, so I've pushed your branch to the main repo to make sure CI is happy before merging. https://github.com/PX4/Firmware/commits/pr-omnibusf4sd

nathantsoi commented:
ahh, thx @davids5. no wonder i couldnt get menuconfig working... copying back in this config works:

```
#
# Board Selection
#
CONFIG_ARCH_BOARD_OMNIBUS_F4SD=y
CONFIG_ARCH_BOARD="omnibus-f4sd"

#
# Custom Board Configuration
#
CONFIG_BOARD_CRASHDUMP=y
CONFIG_BOARD_RESET_ON_CRASH=y
# CONFIG_BOARD_CUSTOM_LEDS is not set
# CONFIG_BOARD_CUSTOM_BUTTONS is not set
CONFIG_BOARD_HAS_PROBES=y
# CONFIG_BOARD_USE_PROBES is not set

#
# Common Board Options
#
CONFIG_ARCH_HAVE_LEDS=y
CONFIG_ARCH_LEDS=y
CONFIG_ARCH_HAVE_BUTTONS=y
CONFIG_ARCH_BUTTONS=y
CONFIG_ARCH_HAVE_IRQBUTTONS=y
CONFIG_ARCH_IRQBUTTONS=y
```

so after running menu config, same thing: switching to uart6 is fine (but then breaks uart6): https://gist.github.com/nathantsoi/5e8ca4ae09151a2fc9fcea64581a73ca

and disabling the serial console (https://gist.github.com/nathantsoi/fbaa1bf116fcebc700b5453251d592e8), i dont get mavlink... this never returns:
```
› Tools/mavlink_shell.py
Using port /dev/serial/by-id/usb-3D_Robotics_OMNIBUSF4SD_0-if00
Connecting to MAVLINK...
```

nathantsoi commented on platforms/nuttx/nuttx-configs/omnibus-f4sd/nsh/defconfig:
```diff
@@ -43,7 +43,7 @@ CONFIG_ARCH_MATH_H=y
 #
 # Debug Options
 #
-CONFIG_DEBUG_ALERT=y
```
@davids5, i went ahead and pushed the results of `make menuconfig` w/o the uart4 changes for now, so we have some baseline for further work

dagar commented on src/modules/sensors/parameters.cpp:
```diff
@@ -188,6 +188,8 @@ void initialize_parameter_handles(ParameterHandles &parameter_handles)
 	(void)param_find("SYS_CAL_TMAX");
 	(void)param_find("SYS_CAL_TMIN");
+
+	(void)param_find("SYS_HAS_BARO");
```
Likewise for SYS_HAS_BARO. As soon as commander starts it runs the preflight checks which finds these.

nathantsoi replied:
done

LorenzMeier commented:
Rebased here: https://github.com/PX4/Firmware/pull/9491

Let's see if CI passes.

LorenzMeier commented:
Merged!
